src/container.ts

```diff
@@ -399,8 +407,29 @@ export function createContainer<T extends object = any, U extends object = any>(
     const keys = [...Object.keys(obj), ...Object.getOwnPropertySymbols(obj)]
 
     for (const key of keys) {
-      const value = obj[key as any] as Resolver<any>
-      registrations[key as any] = value
+      const resolver = obj[key as any] as Resolver<any>
+      // If strict mode is enabled, check to ensure we are not registering a singleton on a non-root
+      // container.
+      if (options?.strict && resolver.lifetime === Lifetime.SINGLETON) {
+        if (parentContainer) {
+          throw new AwilixRegistrationError(
+            key,
+            'Cannot register a singleton on a scoped container.',
+          )
+        }
+      }
+
+      // If this is a value resolver, set the lifetime to `Lifetime.SCOPED` if this is a scoped
+      // container, or `Lifetime.SINGLETON` if this is the root container.
+      if (resolver.isValue) {
```

Comment:
This is what I meant when I was saying I would like to not special case `asValue`. Also, the `aliasTo` isn't covered, right? I should be able to register `aliasTo()` without specifying a lifetime (default being transient) since it'll essentially just "redirect" to the given registration.

My idea with the `isLeakSafe` (again, can workshop the name) was that we wouldn't need to tamper with the lifetime, but rather bypass the check that enforces the case where A depends on B, but A is singleton and B is scoped or transient (B would be _leaking_, since it lives longer than it was supposed to). However, if B is just an `aliasTo`, then it is _safe to leak_ since when resolving the alias, the underlying registration will itself be subjected to the same checks. Or if it's `asValue`, then it's safe to leak because the value itself is already long-lived.

Thoughts?

Reply:
That's an extremely good point. I was worried about `aliasTo` because really it needs to inherit the lifetime value of the registration it resolves to - but this is okay because we carry forward the check in the resolution stack. I was worried that there would be a mismatch between the lifetime defined at resolver creation time and the effective resolution which is undefined because aliasTo can resolve to a registration defined in a lower scope, and that therefore I couldn't give `aliasTo` a meaningful lifetime - but as you point out, we don't actually need to, since it's always a transient that resolves to a registration that has its lifetime checked.

However, I still think there is a problem where A is a singleton and B is a value defined at a lower scope. In this case, we want to detect this lifetime error. The analogous case where A is a singleton and B is a value defined at the same root scope is not an error. That's why values need special behavior that takes into account the container level they are registered at - it's not sufficient to always exclude values from leak checking, the way we should always exclude aliases from leak checking.